debianutils-4.4/tempfile.1
.\" -*- nroff -*-
.TH TEMPFILE 1 "27 Jun 2012" "Debian"
.SH NAME
tempfile \- create a temporary file in a safe manner
.SH SYNOPSIS
.B tempfile
[\-d DIR] [\-p STRING] [\-s STRING] [\-m MODE] [\-n FILE] [\-\-directory=DIR]
[\-\-prefix=STRING] [\-\-suffix=STRING] [\-\-mode=MODE] [\-\-name=FILE] [\-\-help] [\-\-version]
.SH DESCRIPTION
.PP
.B tempfile
creates a temporary file in a safe manner.  It uses
.BR tempnam (3)
to choose the name and opens it with O_RDWR | O_CREAT | O_EXCL.  The filename
is printed on standard output. See
.BR tempnam (3)
for the actual steps involved in directory selection.
.PP
The directory in which to create the file might be searched for in this
order (but refer to
.BR tempnam (3)
for authoritative answers):
.TP 3
a)
In case the environment variable
.B TMPDIR
exists and
contains the name of an appropriate directory, that is used.
.TP
b)
Otherwise, if the
.I \-\-directory
argument is specified and appropriate, it is used.
.TP
c)
Otherwise,
.I P_tmpdir
(as defined in
.IR <stdio.h> )
is used when appropriate.
.TP
d)
Finally an implementation-defined directory
.IR (/tmp)
may be used.
.SH OPTIONS
.TP
.BI "-d, --directory " DIR
Place the file in DIR.
.TP
.BI "-m, --mode " MODE
Open the file with MODE instead of 0600.
.TP
.BI "-n, --name " FILE
Use FILE for the name instead of
.BR tempnam (3) .
The options -d, -p, and -s are ignored if this option is given.
.TP
.BI "-p, --prefix " STRING
Use up to five letters of STRING to generate the name.
.TP
.BI "-s, --suffix " STRING
Generate the file with STRING as the suffix.
.TP
.B "--help"
Print a usage message on standard output and exit successfully.
.TP
.B "--version"
Print version information on standard output and exit successfully.
.SH RETURN VALUES
An exit status of 0 means the temporary file was created successfully.
Any other exit status indicates an error.
.SH BUGS
Exclusive creation is not guaranteed when creating files on NFS
partitions.  tempfile cannot make temporary directories.
tempfile is deprecated; you should use mktemp(1) instead.
.SH EXAMPLE

.RS
.nf
#!/bin/sh
#[...]
t=$(tempfile) || exit
trap "rm -f -- '$t'" EXIT
#[...]
rm -f -- "$t"
trap - EXIT
exit
.fi
.SH "SEE ALSO"
.BR tempnam (3),
.BR mktemp (1)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



debianutils-2.8.4/tempfile.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <getopt.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>

char *progname;

void
usage (int status)
{
  if (status)
    fprintf(stderr, "Try `%s --help' for more information.\n", progname);
  else
    printf("Usage: %s [OPTION]\n\n"
"Create a temporary file in a safe manner.\n\n"
"-d, --directory=DIR  place temporary file in DIR\n"
"-p, --prefix=STRING  set temporary file's prefix to STRING\n"
"-s, --suffix=STRING  set temporary file's suffix to STRING\n"
"-m, --mode=MODE      open with MODE instead of 0600\n"
"-n, --name=FILE      use FILE instead of tempnam(3)\n"
"    --help           display this help and exit\n"
"    --version        output version information and exit\n", progname);
  exit(status);
}


void
syserror (const char *fx)
{
  perror(fx);
  exit(1);
}


int
parsemode (const char *in, mode_t *out)
{
  char *endptr;
  long int mode;
  mode = strtol(in, &endptr, 8);
  if (*endptr || mode<0 || mode>07777)
    return 1;
  *out = (mode_t) mode;
  return 0;
}


int
main (int argc, char **argv)
{
  char *name=0, *dir=0, *pfx=0, *sfx=0;
  mode_t mode = 0600;
  int fd, optc;
  struct option long_options[] = {
    {"prefix", required_argument, 0, 'p'},
    {"suffix", required_argument, 0, 's'},
    {"directory", required_argument, 0, 'd'},
    {"mode", required_argument, 0, 'm'},
    {"name", required_argument, 0, 'n'},
    {"help", no_argument, 0, 'h'},
    {"version", no_argument, 0, 'v'},
    {0, 0, 0, 0}
  };
  progname = argv[0];

  while ((optc = getopt_long (argc, argv, "p:s:d:m:n:", long_options, 0))
	 != EOF) {
    switch (optc) {
    case 0:
      break;
    case 'p':
      pfx = optarg;
      break;
    case 's':
      sfx = optarg;
      break;
    case 'd':
      dir = optarg;
      break;
    case 'm':
      if (parsemode(optarg, &mode)) {
	fprintf(stderr, "Invalid mode `%s'.  Mode must be octal.\n", optarg);
	usage(1);
      }
      break;
    case 'n':
      name = optarg;
      break;
    case 'h':
      usage(0);
    case 'v':
      puts("tempfile " PACKAGE_VERSION);
      exit(0);
    default:
      usage(1);
    }
  }

  if (name) {
    if ((fd = open(name, O_RDWR | O_CREAT | O_EXCL, mode)) < 0)
      syserror("open");
  }

  else {
    for (;;) {
      if (!(name = tempnam(dir, pfx)))
	syserror("tempnam");
      if ((fd = open(name, O_RDWR | O_CREAT | O_EXCL, mode)) < 0) {
	if (errno == EEXIST) {
	  free(name);
	  continue;
	}
	syserror("open");
      }
      if (sfx) {
	char *namesfx;
	if (!(namesfx = malloc(strlen(name) + strlen(sfx) + 1)))
	  syserror("malloc");
	strcpy(namesfx, name);
	strcat(namesfx, sfx);
	if (link(name, namesfx) < 0) {
	  if (errno == EEXIST) {
	    if (unlink(name) < 0)
	      syserror("unlink");
	    free(name);
	    free(namesfx);
	    continue;
	  }
	  syserror("link");
	}
	if (unlink(name) < 0)
	  syserror("unlink");
	free(name);
	name = namesfx;
      }
      break;
    }
  }
  
  if (close(fd))
    syserror("close");
  puts(name);
  exit(0);
}
